File: scripts/apply_migration.py
#!/usr/bin/env python3

"""
CMS Settings Database Migration Script
Applies the CMS settings migration to Supabase PostgreSQL database

Usage: python3 apply_migration.py
"""

import os
import sys
import psycopg2
from psycopg2 import sql
from urllib.parse import urlparse
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).parent.parent / '.env'
load_dotenv(env_path)

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SERVICE_ROLE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
DATABASE_URL = os.getenv('DATABASE_URL')

if not SUPABASE_URL or not SERVICE_ROLE_KEY:
    print("❌ Error: Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env")
    sys.exit(1)

print("╔════════════════════════════════════════════════╗")
print("║  CMS Settings Database Migration Tool          ║")
print("║  Applying: 019_add_cms_settings_tables_FIXED.sql║")
print("╚════════════════════════════════════════════════╝\n")

def get_supabase_connection_string(url: str, service_role_key: str) -> str:
    """
    Extract connection details from Supabase URL and construct PostgreSQL connection string
    Supabase URL format: https://<project-ref>.supabase.co
    """
    # Parse the URL to get project reference
    parsed = urlparse(url)
    project_ref = parsed.netloc.split('.')[0]
    
    # Supabase connection details
    # Default Supabase PostgreSQL port is 5432
    # Database name is typically 'postgres'
    # Username is typically 'postgres'
    
    # Construct connection string
    # Note: You need to use the actual database credentials
    # For Supabase, you typically connect via the pooler or direct connection
    
    conn_str = (
        f"host={parsed.netloc} "
        f"port=5432 "
        f"database=postgres "
        f"user=postgres "
        f"password={service_role_key} "
        f"sslmode=require"
    )
    
    return conn_str

def apply_migration():
    """Execute the migration SQL file"""
    try:
        # Read migration file
        migration_path = Path(__file__).parent.parent / 'migrations' / '002-core-features' / '019_add_cms_settings_tables_FIXED.sql'
        
        if not migration_path.exists():
            print(f"❌ Error: Migration file not found at {migration_path}")
            sys.exit(1)
        
        with open(migration_path, 'r') as f:
            sql_content = f.read()
        
        print("📝 Migration file loaded")
        print(f"📦 Target: {SUPABASE_URL}\n")
        print("⏳ Executing migration...\n")
        
        # Try to connect to the database
        try:
            conn_str = get_supabase_connection_string(SUPABASE_URL, SERVICE_ROLE_KEY)
            conn = psycopg2.connect(conn_str)
        except psycopg2.Error as e:
            print(f"⚠ Could not connect via direct PostgreSQL: {e}")
            print("ℹ Trying alternative connection method...\n")
            
            # If direct connection fails, you might need to:
            # 1. Use Supabase's SQL Editor web interface
            # 2. Use psql CLI with proper credentials
            # 3. Use Supabase's API with proper authentication
            
            print("❌ Direct PostgreSQL connection failed.")
            print("\n📋 Alternative methods to apply this migration:\n")
            print("1️⃣  Use Supabase SQL Editor (Web UI):")
            print(f"   - Go to: {SUPABASE_URL}/project/sql/new")
            print("   - Copy the SQL from: Backend/migrations/002-core-features/019_add_cms_settings_tables_FIXED.sql")
            print("   - Paste into the editor and execute\n")
            
            print("2️⃣  Use Supabase CLI:")
            print("   - Install: npm install -g supabase")
            print("   - Run: supabase db pull")
            print("   - Then apply migrations\n")
            
            print("3️⃣  Use psql command line:")
            print(f"   - Get your PostgreSQL credentials from Supabase dashboard")
            print(f"   - Run: psql -h <host> -U postgres -f Backend/migrations/002-core-features/019_add_cms_settings_tables_FIXED.sql\n")
            
            sys.exit(1)
        
        # Execute the migration
        cursor = conn.cursor()
        
        # Split into statements and execute
        statements = [
            stmt.strip() 
            for stmt in sql_content.split(';') 
            if stmt.strip() and not stmt.strip().startswith('--')
        ]
        
        print(f"Found {len(statements)} SQL statements\n")
        
        executed = 0
        for i, statement in enumerate(statements, 1):
            try:
                cursor.execute(statement)
                executed += 1
                print(f"✓ Statement {i}/{len(statements)} executed")
            except psycopg2.Error as e:
                # Some errors are expected (IF NOT EXISTS)
                if 'already exists' in str(e) or 'does not exist' in str(e):
                    print(f"⚠ Statement {i}/{len(statements)} skipped (object already exists)")
                    executed += 1
                else:
                    print(f"✗ Statement {i}/{len(statements)} failed: {e}")
                    conn.rollback()
                    raise
        
        conn.commit()
        cursor.close()
        conn.close()
        
        print(f"\n✅ Migration completed! {executed}/{len(statements)} statements executed")
        print("\n📋 CMS Settings tables created successfully!")
        print("   - cms_settings (with settings_data JSONB column)")
        print("   - cms_settings_versions")
        print("   - cms_audit_log")
        print("   - theme_presets (with 6 default themes)")
        
    except Exception as e:
        print(f"\n❌ Migration failed: {e}")
        sys.exit(1)

if __name__ == '__main__':
    apply_migration()

File: scripts/apply-migration.js
#!/usr/bin/env node

/**
 * Migration Application Script
 * Applies the CMS Settings migration to Supabase database
 * Usage: node apply-migration.js
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// Load environment variables
require('dotenv').config({ path: path.join(__dirname, '../.env') });

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SERVICE_ROLE_KEY) {
  console.error('❌ Error: Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in .env');
  process.exit(1);
}

// Parse Supabase URL to get host
const url = new URL(SUPABASE_URL);
const host = url.hostname;

/**
 * Execute SQL query against Supabase REST API
 * Using the /rest/v1/rpc/execute_sql endpoint
 */
async function executeSql(sql) {
  return new Promise((resolve, reject) => {
    const postData = JSON.stringify({ query: sql });

    const options = {
      hostname: host,
      port: 443,
      path: '/rest/v1/rpc/sql_execute',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${SERVICE_ROLE_KEY}`,
        'apikey': SERVICE_ROLE_KEY,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
      },
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve({ status: res.statusCode, data });
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', (e) => {
      reject(e);
    });

    req.write(postData);
    req.end();
  });
}

/**
 * Execute migration from SQL file
 */
async function applyMigration() {
  try {
    console.log('🔄 Starting CMS Settings migration...\n');

    // Read migration file
    const migrationPath = path.join(__dirname, '../migrations/002-core-features/019_add_cms_settings_tables_FIXED.sql');
    const sql = fs.readFileSync(migrationPath, 'utf8');

    console.log('📝 Migration file loaded');
    console.log(`📦 Target: ${SUPABASE_URL}\n`);

    // Execute migration
    console.log('⏳ Executing migration SQL...');
    
    // Split the SQL into individual statements for better error handling
    const statements = sql
      .split(';')
      .map(stmt => stmt.trim())
      .filter(stmt => stmt.length > 0 && !stmt.startsWith('--'));

    console.log(`📊 Found ${statements.length} SQL statements to execute\n`);

    let executed = 0;
    for (const statement of statements) {
      try {
        await executeSql(statement);
        executed++;
        console.log(`✓ Statement ${executed} executed`);
      } catch (error) {
        // Some statements might fail due to existing objects, which is expected
        if (error.message.includes('already exists') || error.message.includes('does not exist')) {
          console.log(`⚠ Statement ${executed} skipped (${error.message.substring(0, 50)}...)`);
          executed++;
        } else {
          throw error;
        }
      }
    }

    console.log(`\n✅ Migration completed! ${executed}/${statements.length} statements executed`);
    console.log('\n📋 CMS Settings tables should now have the settings_data JSONB column');

  } catch (error) {
    console.error('\n❌ Migration failed:');
    console.error(error.message);
    process.exit(1);
  }
}

// Alternative approach: Use psql if available
async function applyMigrationViaRpc() {
  try {
    console.log('🔄 Starting CMS Settings migration via Supabase RPC...\n');

    // Read migration file
    const migrationPath = path.join(__dirname, '../migrations/002-core-features/019_add_cms_settings_tables_FIXED.sql');
    const sql = fs.readFileSync(migrationPath, 'utf8');

    console.log('📝 Migration file loaded');
    console.log(`📦 Target: ${SUPABASE_URL}\n`);

    // Use Supabase's sql() RPC method
    console.log('⏳ Executing migration SQL via Supabase RPC...');

    const postData = JSON.stringify({
      query: sql
    });

    const options = {
      hostname: host,
      port: 443,
      path: '/rest/v1/rpc/migrations',
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${SERVICE_ROLE_KEY}`,
        'apikey': SERVICE_ROLE_KEY,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
      },
    };

    return new Promise((resolve, reject) => {
      const req = https.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            console.log('\n✅ Migration completed successfully!');
            console.log('\n📋 CMS Settings tables created with settings_data JSONB column');
            resolve(data);
          } else {
            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          }
        });
      });

      req.on('error', reject);
      req.write(postData);
      req.end();
    });

  } catch (error) {
    console.error('\n❌ Migration failed:');
    console.error(error.message);
    process.exit(1);
  }
}

// Run the migration
console.log('╔════════════════════════════════════════════════╗');
console.log('║  CMS Settings Database Migration Tool          ║');
console.log('║  Applying: 019_add_cms_settings_tables_FIXED.sql║');
console.log('╚════════════════════════════════════════════════╝\n');

// Try the simpler RPC approach first
applyMigrationViaRpc().catch(() => {
  console.log('\nℹ Retrying with alternative method...\n');
  applyMigration();
});
